File: Top-Interview-Questions/Array/Easy-Collection/136-single_number.cpp
class Solution {
public:
    int singleNumber(vector<int>& nums) {
        /*
         * Time Complexity: O(n)
         * Space Complexity: O(n)
         * 
         */
        int n = nums.size();
        
        if(n == 1){
            return nums[0];
        }
        
        map<int, bool> a;
        int sum = 0;
        
        for(int i = 0; i < n; i++){
            if(a[nums[i]] == false){
                sum += nums[i];
                a[nums[i]] = true;
            }else{
                sum -= nums[i];
            }
        }
        
        return sum;
        
    }
};

class Solution {
public:
    int singleNumber(vector<int>& nums) {
        
        /*
         * Time Complexity: O(n^2)
         * Space Complexity: O(1)
         * Time Limit Exceeded
         * Reason: Uses no memory but time complexity is too high
         * 
         */
        
        int n = nums.size();
        int result;
        
        bool check = false;
        if(n == 1){
            return nums[0];
        }
                
        for(int i = 0; i < n; i++){
            check = false;
            
            for(int j = 0; j < n; j++){
                if(j != i){
                    if(nums[i] == nums[j]){
                        check = true;
                    }
                }
                
                if(j == n-1){
                    if(check == false){
                        result = nums[i];
                        return result;
                    }
                }
            }
        }
        
        return result;
    }
};

File: Top-Interview-Questions/Array/Easy-Collection/136-single-number.py
class Solution:
    # Sol 1. List operation
    def singleNumber(self, nums: List[int]) -> int:
        new_list = []
        for i in nums:
            if i not in new_list:
                new_list.append(i)
            else:
                new_list.remove(i)
        return new_list.pop()
    # Time Complexity: O(n) 
    # Space complexity: O(n)

    # Sol 2. Math
    '''
    Each character can be duplicated up to 2 times.
    2*(a+b+c)-(a+a+b+b+c) = c
    '''
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        return 2 * sum(set(nums)) - sum(nums)
    # Time Complexity = O(n+n) = O(n)
    # Space Complexity = O(n)

    # Sol 3. Bit Manipulation, consider this as binary numbers
    # XOR, in python 3, it is described as ^ operator
    # a xor a = 0, a xor 0 = a
    def singleNumber(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        a = 0
        for i in nums:
            a ^= i
        return a

    # Time Complexity: O(n)
    # Space Complexity: O(1)

    # Sol 4. Using Count
    class Solution:
        def singleNumber(self, nums: List[int]) -> int:
            for i in nums:
                if nums.count(i) == 1:
                    return i
    # Time Complexity: O(n)
    # Space Complexity: O(1)
